File: drone_forest/forest.py
"""Script for the forest class."""

import numpy as np
from typing import List, Tuple

from drone_forest.tree import Tree
from drone_forest.geometric_objects import Circle, Point
from drone_forest.geometric_utils import distance


class Forest:
    """A forest object in the drone forest.

    Attributes:
        trees (List[Tree]): The trees in the forest.
    """

    def __init__(
        self,
        xlim: Tuple[float, float],
        ylim: Tuple[float, float],
        n_trees: int,
        max_radius: float,
        exclusion_zones: List[Circle] = [],
        min_spare_distance: float = 0.2,
        max_spawn_attempts: int = 50,
    ):
        """Initialize the forest object.

        Args:
            trees (List[Tree]): The trees in the forest.
        """
        assert max_radius >= 0, "The maximum radius of the trees must be non-negative."
        assert n_trees > 0, "The number of trees must be positive."
        assert (
            min_spare_distance >= 0
        ), "The minimum spare distance must be non-negative."
        assert max_spawn_attempts > 0, "The maximum spawn attempts must be positive."
        assert xlim[0] < xlim[1], "The x-axis limits must be in increasing order."
        assert ylim[0] < ylim[1], "The y-axis limits must be in increasing order."

        self.trees: List[Tree] = []
        self.xlim = xlim
        self.ylim = ylim

        # Generate random trees
        for _ in range(n_trees):
            # Try to spawn a tree
            for _ in range(max_spawn_attempts):
                x_pos = np.random.uniform(xlim[0], xlim[1])
                y_pos = np.random.uniform(ylim[0], ylim[1])
                radius = np.random.uniform(
                    0.05,
                    min(
                        max_radius,
                        xlim[1] - x_pos,
                        x_pos - xlim[0],
                        ylim[1] - y_pos,
                        y_pos - ylim[0],
                    ),
                )

                # Check if the tree is too close to another tree
                if any(
                    distance(Point(x_pos, y_pos), tree.get_position())
                    < radius + tree.get_radius() + min_spare_distance
                    for tree in self.trees
                ):
                    continue

                # Check if the tree is not in the exclusion zone
                if any(
                    distance(Point(x_pos, y_pos), circle.center)
                    < radius + circle.radius
                    for circle in exclusion_zones
                ):
                    continue

                self.trees.append(Tree(Point(x_pos, y_pos), radius))
                break

    def draw(self, ax):
        """Draw the forest on the given axis.

        Args:
            ax (matplotlib.axes.Axes): The axis to draw the forest on.
        """
        assert ax is not None, "The axis must be provided."

        for tree in self.trees:
            tree.draw(ax)

File: src/forest.cc
#include <drone_forest/forest.h>

namespace evs
{
namespace forest
{

std::mt19937 Forest::gen_(std::random_device{}());

Forest::Forest(const std::tuple<double, double> x_limits,
               const std::tuple<double, double> y_limits, int num_trees,
               double min_radius, double max_radius,
               std::vector<geometric::Circle> exclusion_zones,
               double min_spare_distance, int max_spawn_attempts)
{
  // Generate uniform distribution for x and y
  std::uniform_real_distribution<double> x_dist(std::get<0>(x_limits),
                                                std::get<1>(x_limits));
  std::uniform_real_distribution<double> y_dist(std::get<0>(y_limits),
                                                std::get<1>(y_limits));
  std::uniform_real_distribution<double> radius_dist(min_radius, max_radius);

  // Generate trees
  for (int idx_tree = 0; idx_tree < num_trees; idx_tree++)
  {
    for (int idx_attempt = 0; idx_attempt < max_spawn_attempts; idx_attempt++)
    {
      // Generate random tree
      Tree tree(geometric::Point(x_dist(gen_), y_dist(gen_)),
                radius_dist(gen_));

      // Check if tree is in exclusion zone
      bool in_exclusion_zone = false;
      for (const geometric::Circle& exclusion_zone : exclusion_zones)
      {
        if (tree.Trunk().Center().Distance(exclusion_zone.Center())
            < tree.Trunk().Radius() + exclusion_zone.Radius())
        {
          in_exclusion_zone = true;
          break;
        }
      }

      // Skip tree if in exclusion zone
      if (in_exclusion_zone)
      {
        continue;
      }

      // Check if minimum distance to other trees is satisfied
      bool min_distance_satisfied = true;
      for (const Tree& other_tree : trees_)
      {
        if (tree.Trunk().Center().Distance(other_tree.Trunk().Center())
            < tree.Trunk().Radius() + other_tree.Trunk().Radius()
                  + min_spare_distance)
        {
          min_distance_satisfied = false;
          break;
        }
      }

      // Skip tree if minimum distance is not satisfied
      if (!min_distance_satisfied)
      {
        continue;
      }

      // Add tree to forest
      trees_.push_back(Tree(tree));
      break;
    }
  }
}

void Forest::Draw(cv::Mat& img, geometric::Point t_vec, double m2px) const
{
  for (const Tree& tree : trees_)
  {
    tree.Draw(img, t_vec, m2px);
  }
}

std::vector<geometric::Circle> Forest::GetObstacles() const
{
  std::vector<geometric::Circle> obstacles;
  for (const Tree& tree : trees_)
  {
    obstacles.push_back(tree.Trunk());
  }
  return obstacles;
}

int Forest::NumberOfTrees() const
{
  return trees_.size();
}

std::vector<Tree> Forest::Trees() const
{
  return trees_;
}

}  // namespace forest
}  // namespace evs